File: unsorted/js/76.minimum-window-substring.js
/*
 * @lc app=leetcode id=76 lang=javascript
 *
 * [76] Minimum Window Substring
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {string}
 */
var minWindow = function(s, t) {
	let valids = t.split('');
	let hash = new Map();
	valids.forEach((v) => {
		hash.set(v, 0);
	});

	let isValid = function(hash) {
		let flag = true;
		hash.forEach((value, key) => {
			if (value === 0) flag = false;
			//    console.log('value', value);
			//     console.log('flag', flag);
		});
		// console.log('flag', flag);
		return flag;
	};

	let i = 0,
		j = 0;
	let min = Number.MAX_VALUE;
	let minStr = '';

	for (; i < s.length; i++) {
		while (j < s.length) {
			// console.log('i',i,'j',j);
			// console.log('hash', hash);
			// console.log('str', s.slice(i, j+1));
			const ch = s[j];
			if (hash.has(ch)) {
				hash.set(ch, 1);
			}

			if (isValid(hash)) {
				if (j - i + 1 < min) {
					min = j - 1 + 1;
					minStr = s.substr(i, j + 1);
				}
				break;
			}
			else {
				j++;
			}
		}
		if (hash.has(s[i])) {
			hash.set(s[i], 0);
		}
	}
	return minStr;
};
// @lc code=end

File: unsorted/java/76.minimum-window-substring.java
/*
 * @lc app=leetcode id=76 lang=java
 *
 * [76] Minimum Window Substring
 */

// @lc code=start
class Solution {


    public String minWindow(String s, String t) {

        Map<Character, Integer> tMap = new HashMap<>();
        Map<Character, Integer> sMap = new HashMap<>();
        for (char a : t.toCharArray()) {
            if (!tMap.containsKey(a)) {
                tMap.put(a, 1);
            } else {
                tMap.put(a, tMap.get(a) + 1);
            }
        }
        for (char a : s.toCharArray()) {
            sMap.put(a, 0);
        }
        int l = 0, r = -1;
        boolean found = false;
        String res = s;
        while (l < s.length()) {
            if (r + 1 < s.length() && !tMapFilled(sMap, tMap)) {
                r++;
                char cr = s.charAt(r);
                sMap.put(cr, sMap.get(cr) + 1);
            } else {
                char cl = s.charAt(l);
                sMap.put(cl, sMap.get(cl) - 1);
                l++;
            }

            if (tMapFilled(sMap, tMap) && r - l + 1 <= res.length()) {
                found = true;
                res = s.substring(l, r + 1);
                // System.out.println("answer " +res);

            }
        }
        if (found == false) {
            return "";
        } else {
            return res;
        }

    }

    boolean tMapFilled(Map<Character, Integer> sMap, Map<Character, Integer> tMap) {
        for (Character a : tMap.keySet()) {
            if (!sMap.containsKey(a) || sMap.get(a) < tMap.get(a)) {
                return false;
            }
        }
        return true;
    }
}
// @lc code=end
